code/go/0chain.net/core/util/merkle_patricia_trie.go

```diff
@@ -129,14 +129,15 @@ func (mpt *MerklePatriciaTrie) Insert(path Path, value Serializable) (Key, error
 		return mpt.Delete(path)
 	}
 
+	valueCopy := &SecureSerializableValue{eval}
```

Comment:
as `wallet_test` failed because it saved an updated `State` object multiple times into MPT, an as MemoryNodeDB actually stores original objects in its map, subsequent MPT Validation failed because some nodes hashes were recalculated on AddChanges into different values.
So the safest solution is to convert any incoming value to bytes (SecureSerializableValue) right on input to the MPT.

Reply:
LGTM 👍 . This is the minimal changes we can do to make it immutable, but in the future, we will stop doing the `encoding` in the `Insert()` method of MPT, the encoding should happen only when we save the data into persistent db (i.e. *pnode), all other types of nodes store value in RAM directly.

Reply:
@peterlimg it will be extremely useful to extract event sequence (updates, deletes, inserts, reads) before MPT persistence to track changes caused by transaction execution. it is the cornerstone of parallelizing smart contract execution which I'm working on now. I will describe how it can help parallelization soon in details, may be you thought of addition such a functionality previously?

Reply:
Changes are currently collected in ChangeCollector object (each MPT has it), if that's what you're asking about.
The question is how to merge changes from several collectors and see if they really conflict or not. Current implementation doesn't allow to do this.

Reply:
Yes, something similar to ChangeCollector with additional AccessCollector. We will need some mechanics which can tell us what parts of the state SC accessed and what changed to make read and write sets for further contention detection.

Reply:
sub-trees, that is?